docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,264 @@
+
+**What's good about this** is that an "ffi module" (and there may be many) defines the canonical API
+and it is defined in terms of Rust types - the redundant UDL is removed.
+The Rust scaffolding can also be generated by the macros, meaning there are no generated `.rs`
+files involved. Types can be shared among any of the ffi modules defined in the project -
+for example, [this diplomat ffi module](https://github.com/unicode-org/icu4x/blob/7d9f89fcd7df4567e17ddd8c46810b0db287436a/ffi/diplomat/src/pluralrules.rs#L50-L51)
+uses types from a [different ffi module](https://github.com/unicode-org/icu4x/blob/7d9f89fcd7df4567e17ddd8c46810b0db287436a/ffi/diplomat/src/locale.rs#L19).
+
+It also offers better control over the stability of the API, because where the FFI is defined
+is constrained. This is [an explicit design decision](https://github.com/rust-diplomat/diplomat/blob/main/docs/design_doc.md#requirements) of diplomat.
+
+While the process for defining the type universe is different, the actual in-memory
+representation of that type universe isn't radically different from UniFFI - for example,
+here's the [definition of a Rust struct](https://github.com/rust-diplomat/diplomat/blob/main/core/src/ast/structs.rs),
+and while it is built from a `syn` struct, the final representation is independent of `syn`
+and its ast representation.
+
+## UniFFI's experience with the macro approach.
+
+Ryan tried this same macro approach for UniFFI in [#416](https://github.com/mozilla/uniffi-rs/pull/416) -
+but we struck **what's bad about this** - at least for UniFFI's use-cases: the context in which the
+macro runs doesn't know about types defined outside of that macro, which are what we need to
+expose.
+
+### Example of these limitations
+
+Let's look at diplomat's simple example:
+
+```rust
+#[diplomat::bridge]
+mod ffi {
+    pub struct MyFFIType {
+        pub a: i32,
+        pub b: bool,
+    }
+
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+This works fine, but starts to come unstuck if you want the types defined somewhere else. In this trivial example, something like:
+
+```Rust
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+
+#[diplomat::bridge]
+mod ffi {
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+fails - diplomat can't handle this scenario - in the same way and for the same reasons that Ryan's
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) can't - the contents of the struct aren't known.
+
+From the Rust side of the world, this is probably solvable by sprinkling more macros around - eg, something like:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+Might be enough for the generation of the Rust scaffolding. However, the problems are in the
+foreign bindings, because, eg, those foreign bindings do not know the names and types of the
+struct elements.
```

Comment:
Can you elaborate on this one?  If we can use these macros as the source for scaffolding code, why couldn't we generate the bindings code as well?

Reply:
I hope I cleared this up in a86e9496623f9bbee77fb7b3b23e960ee47ebdc2 - lmk if I need more about this.